util/dvsim/NcLauncher.py

```diff
@@ -150,14 +163,44 @@ def poll(self):
         """
 
         assert self.process is not None
-        elapsed_time = datetime.datetime.now() - self.start_time
-        job_runtime_secs = elapsed_time.total_seconds()
         if self.process.poll() is None:
-            timeout_mins = self.deploy.get_timeout_mins()
-            if timeout_mins is not None and not self.deploy.gui:
-                if job_runtime_secs > (timeout_mins * 60):
+            run_timeout_mins = self.deploy.get_timeout_mins()
+            if run_timeout_mins is not None and not self.deploy.gui:
+                # We consider the job to have started once its log file contains
+                # something. file_size_thresh_bytes is a threshold: once the log
+                # file is bigger than this many bytes, the job must have started
+                wait_timeout_mins = 180  # max wait time in job / license queue
+                file_size_thresh_bytes = 5120  # log file size threshold
+
+                # query the log file size
+                f_size = os.path.getsize(self.deploy.get_log_path())
+
+                if f_size >= file_size_thresh_bytes:
+                    if (self.nc_job_state == 'waiting'):
+                        # If the job log size is more than the threshold,
+                        # declare the job to have started running
+                        # capture the run start time
+                        self.start_time = datetime.datetime.now()
+                        self.nc_job_state = 'running'
+
+                if self.nc_job_state == 'waiting':
+                    # If we get here, we know the log size is less than the threshold.
+                    # check if wait_timeout_mins has elapsed.
+                    if (self.minutes_since_start() > wait_timeout_mins):
+                        self.nc_job_state = 'wait_timeout'
+
+                if self.nc_job_state == 'running':
+                    if (self.minutes_since_start() > run_timeout_mins):
+                        self.nc_job_state = 'run_timeout'
+
+                if self.nc_job_state in ('wait_timeout', 'run_timeout'):
```

Comment:
Not a big deal, but I'd suggest using a list or set to avoid a reader thinking about membership of tuples. I just checked and this code does work (a surprise to me!) but it's not super obvious. I'd probably suggest `{ ... }`.

Reply:
done